packages/kbn-management/settings/components/form/form.tsx

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React, { Fragment } from 'react';
+
+import type { FieldDefinition } from '@kbn/management-settings-types';
+import { FieldRow, RowOnChangeFn } from '@kbn/management-settings-components-field-row';
+import { SettingType, UnsavedFieldChange } from '@kbn/management-settings-types';
+import { isEmpty } from 'lodash';
+import { i18n } from '@kbn/i18n';
+import { BottomBar } from './bottom_bar';
+import { useServices } from './services';
+
+/**
+ * Props for a {@link Form} component.
+ */
+export interface FormProps {
+  /** A list of {@link FieldDefinition} corresponding to settings to be displayed in the form. */
+  fields: Array<FieldDefinition<SettingType>>;
+  /** True if saving settings is enabled, false otherwise. */
+  isSavingEnabled: boolean;
+}
+
+/**
+ * Component for displaying a set of {@link FieldRow} in a form.
+ * @param props The {@link FormProps} for the {@link Form} component.
+ */
+export const Form = (props: FormProps) => {
+  const { fields, isSavingEnabled } = props;
+  const { saveChanges, showError, showReloadPagePrompt } = useServices();
+
+  const [unsavedChanges, setUnsavedChanges] = React.useState<
+    Record<string, UnsavedFieldChange<SettingType>>
+  >({});
+
+  const [isLoading, setIsLoading] = React.useState<boolean>(false);
+
+  const unsavedChangesCount = Object.keys(unsavedChanges).length;
+  const hasInvalidChanges = Object.values(unsavedChanges).some(({ isInvalid }) => isInvalid);
+
+  const saveAll = async () => {
+    setIsLoading(true);
+    if (isEmpty(unsavedChanges)) {
```

Comment:
nit: you might consider abstracting this into a hook.  It likely won't be reused, but it's cleaner/more clear.  YMMV

Reply:
This is a good idea! It also makes sense to abstract it into a hook because all service functions are only used in this logic so we can just call `useServices` inside this new hook rather than in the form component.

I added the hook in a separate file inside the same package - I was thinking of adding it to the `utilities` package but not sure if it's worth adding a whole `form` folder there just for this hook, but let me know if you think otherwise.